File: backend-api/src/main/java/pl/com/chrzanowski/sma/scaffolding/position/service/filter/ScaffoldingLogPositionFilter.java
package pl.com.chrzanowski.sma.scaffolding.position.service.filter;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import pl.com.chrzanowski.sma.common.enumeration.ScaffoldingType;
import pl.com.chrzanowski.sma.common.enumeration.TechnicalProtocolStatus;

import java.math.BigDecimal;
import java.time.LocalDate;

@Builder(toBuilder = true)
@AllArgsConstructor
@NoArgsConstructor(force = true)
@Data
public class ScaffoldingLogPositionFilter {

    private Long id;
    private String scaffoldingNumberContains;
    private String assemblyLocationContains;
    private LocalDate assemblyDateGreaterOrEqual;
    private LocalDate assemblyDateLessOrEqual;
    private LocalDate dismantlingDateGreaterOrEqual;
    private LocalDate dismantlingDateLessOrEqual;
    private LocalDate dismantlingNotificationDateGreaterOrEqual;
    private LocalDate dismantlingNotificationDateLessOrEqual;
    private TechnicalProtocolStatus technicalProtocolStatus;
    private ScaffoldingType scaffoldingType;
    private BigDecimal scaffoldingFullDimensionGreaterOrEqual;
    private BigDecimal scaffoldingFullDimensionLessOrEqual;
    private String scaffoldingLogNameContains;
    private String scaffoldingLogAdditionalInfoContains;
    private String contractorNameContains;
    private String contractorContactNameContains;
    private String scaffoldingUserNameContains;
    private String scaffoldingUserContactNameContains;

    //todo in future extend filter of working time (qoantuty of hours multiply quantity of workers) and dimensions maybe

}

File: frontend/react/src/filters/scaffolding-log-position-filter.ts
import {TechnicalProtocolStatus} from "@/enums/technical-protocol-status-types-enum.ts";
import {ScaffoldingType} from "@/enums/scaffolding-type-types-enum.ts";

export interface ScaffoldingLogPositionFilter {
    id?: number;
    scaffoldingNumberContains?: string;
    assemblyLocationContains?: string;
    assemblyDateGreaterOrEqual?: string;
    assemblyDateLessOrEqual?: string;
    dismantlingDateGreaterOrEqual?: string;
    dismantlingDateLessOrEqual?: string;
    dismantlingNotificationDateGreaterOrEqual?: string;
    dismantlingNotificationDateLessOrEqual?: string;
    technicalProtocolStatus?: TechnicalProtocolStatus;
    scaffoldingType?: ScaffoldingType;
    scaffoldingFullDimensionGreaterOrEqual?: string;
    scaffoldingFullDimensionLessOrEqual?: string;
    scaffoldingLogNameContains?: string;
    scaffoldingLogAdditionalInfoContains?: string;
    contractorNameContains?: string;
    contractorContactNameContains?: string;
    scaffoldingUserNameContains?: string;
    scaffoldingUserContactNameContains?: string;
    parentPositionOnly?: boolean;

    page?: number;
    size?: number;
    sort?: string;
}

export class ScaffoldingLogPositionFilterBuilder {
    private readonly f: Partial<ScaffoldingLogPositionFilter> = {};

    withId(id: number) {
        this.f.id = id;
        return this;
    }

    withScaffoldingNumberContains(scaffoldingNumberContains: string) {
        this.f.scaffoldingNumberContains = scaffoldingNumberContains;
        return this;
    }

    withAssemblyLocationContains(assemblyLocationContains: string) {
        this.f.assemblyLocationContains = assemblyLocationContains;
        return this;
    }

    withAssemblyDateGreaterOrEqual(date: string) {
        this.f.assemblyDateGreaterOrEqual = date;
        return this;
    }

    withAssemblyDateLessOrEqual(date: string) {
        this.f.assemblyDateLessOrEqual = date;
        return this;
    }

    withDismantlingDateGreaterOrEqual(date: string) {
        this.f.dismantlingDateGreaterOrEqual = date;
        return this;
    }

    withDismantlingDateLessOrEqual(date: string) {
        this.f.dismantlingDateLessOrEqual = date;
        return this;
    }

    withDismantlingNotificationDateGreaterOrEqual(date: string) {
        this.f.dismantlingNotificationDateGreaterOrEqual = date;
        return this;
    }

    withDismantlingNotificationDateLessOrEqual(date: string) {
        this.f.dismantlingNotificationDateLessOrEqual = date;
        return this;
    }

    withTechnicalProtocolStatus(status: TechnicalProtocolStatus) {
        this.f.technicalProtocolStatus = status;
        return this;
    }

    withScaffoldingType(type: ScaffoldingType) {
        this.f.scaffoldingType = type;
        return this;
    }

    withScaffoldingFullDimensionGreaterOrEqual(dimension: string) {
        this.f.scaffoldingFullDimensionGreaterOrEqual = dimension;
        return this;
    }

    withScaffoldingFullDimensionLessOrEqual(dimension: string) {
        this.f.scaffoldingFullDimensionLessOrEqual = dimension;
        return this;
    }

    withScaffoldingLogNameContains(name: string) {
        this.f.scaffoldingLogNameContains = name;
        return this;
    }

    withScaffoldingLogAdditionalInfoContains(info: string) {
        this.f.scaffoldingLogAdditionalInfoContains = info;
        return this;
    }

    withContractorNameContains(name: string) {
        this.f.contractorNameContains = name;
        return this;
    }

    withContractorContactNameContains(name: string) {
        this.f.contractorContactNameContains = name;
        return this;
    }

    withScaffoldingUserNameContains(name: string) {
        this.f.scaffoldingUserNameContains = name;
        return this;
    }

    withScaffoldingUserContactNameContains(name: string) {
        this.f.scaffoldingUserContactNameContains = name;
        return this;
    }

    withParentPositionOnly(isParent: boolean) {
        this.f.parentPositionOnly = isParent;
        return this;
    }

    withPage(page: number) {
        this.f.page = page;
        return this;
    }

    withSize(size: number) {
        this.f.size = size;
        return this;
    }

    withSort(sort: string) {
        this.f.sort = sort;
        return this;
    }

    build(): ScaffoldingLogPositionFilter {
        return {...this.f};
    }
}
